device.go

```diff
@@ -361,21 +361,42 @@ func (d *Device) foundCandidate(nonce1 uint32, nonce0 uint32) {
 	copy(data, d.work.Data[:])
 	binary.BigEndian.PutUint32(data[128+4*nonce1Word:], nonce1)
 	binary.BigEndian.PutUint32(data[128+4*nonce0Word:], nonce0)
-	hash := chainhash.HashFuncB(data[0:180])
 
-	newHash, err := chainhash.NewHashFromStr(hex.EncodeToString(reverse(hash[:])))
-	if err != nil {
-		minrLog.Error(err)
-	}
-	minrLog.Errorf("hash: %x", hash)
-	minrLog.Errorf("newHash: %v", newHash)
-	hashNum := blockchain.ShaHashToBig(newHash)
-	if hashNum.Cmp(d.work.Target) > 0 {
-		minrLog.Infof("Hash %s below target %s", hex.EncodeToString(reverse(hash[:])), d.work.Target)
+	// XXX Really we should be able to use the same test in both cases
+	// but for now in case there are some idiosyncrasies in the original
+	// keep it for solo (testnet) mining.
+	if cfg.Pool != "" {
+		hash := chainhash.HashFuncB(data[0:180])
 
+		newHash, err := chainhash.NewHashFromStr(hex.EncodeToString(reverse(hash[:])))
+		if err != nil {
+			minrLog.Error(err)
+		}
+		minrLog.Errorf("hash: %x", hash)
+		minrLog.Errorf("newHash: %v", newHash)
+		hashNum := blockchain.ShaHashToBig(newHash)
+		if hashNum.Cmp(d.work.Target) > 0 {
+			minrLog.Infof("Hash %s below target %s", hex.EncodeToString(reverse(hash[:])), d.work.Target)
+
+		} else {
+			minrLog.Infof("Found hash!!  %s", hex.EncodeToString(hash[:]))
+			d.workDone <- data
+		}
 	} else {
-		minrLog.Infof("Found hash!!  %s", hex.EncodeToString(hash[:]))
-		d.workDone <- data
+		// Perform the final hash block to get the hash
+		var state [8]uint32
+		copy(state[:], d.midstate[:])
+		blake256.Block(state[:], data[128:192], 1440)
+
+		var hash [32]byte
+		for i := 0; i < 8; i++ {
+			binary.BigEndian.PutUint32(hash[i*4:], state[i])
+		}
+
+		if hashSmaller(hash[:], d.work.Target[:]) {
```

Comment:
This doesn't compile (d.work.Target[:] should be d.work.Target.Bytes()).  With that change the code panics in hashSmaller().  Reversing the args lets it run but I don't think that's correct.